streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,283 @@
+        ByteType: 'bytes',
+        ShortType: 'uint64',
+        IntegerType: 'int',
+        LongType: 'int64',
+        FloatType: 'float32',
+        DoubleType: 'float64',
+        DecimalType: 'str_decimal',
+        StringType: 'str',
+        BinaryType: 'bytes',
+        BooleanType: None,
+        TimestampType: None,
+        TimestampNTZType: None,
+        DateType: None,
+        DayTimeIntervalType: None,
+        ArrayType: None,
+        MapType: None,
+        StructType: None,
+        StructField: None
+    }
+
+    def map_spark_dtype(spark_data_type: Any):
+        for spark_dtype, mds_dtype in mapping_spark_to_mds.items():
+            if isinstance(spark_data_type, spark_dtype) and mds_dtype is not None:
+                return mds_dtype
+        raise ValueError(f'{spark_data_type} is not supported by MDSwriter')
+
+    if user_defined_cols is not None:  # user has provided schema, we just check if mds supports the dtype
+        mds_supported_dtypes = set(mapping_spark_to_mds.values())
+        for col_name, user_dtype in user_defined_cols.items():
+            if col_name not in dataframe.columns:
+                raise ValueError(
+                    f'{col_name} is not a column of input dataframe: {dataframe.columns}')
+            if user_dtype not in mds_supported_dtypes:
+                raise ValueError(f'{user_dtype} is not supported by MDSwriter')
+
+            actual_spark_dtype = dataframe.schema[col_name].dataType
+            mapped_mds_dtype = map_spark_dtype(actual_spark_dtype)
+            if user_dtype != mapped_mds_dtype:
+                raise ValueError(f"Mismatched types: {col_name} is {mapped_mds_dtype} in DataFrame but {user_dtype} in user_defined_cols")
+        return None
+
+    schema = dataframe.schema
+    schema_dict = {}
+
+    for field in schema:
+        dtype = map_spark_dtype(field.dataType)
+        if dtype in _encodings:
+            schema_dict[field.name] = dtype
+        else:
+            raise ValueError(f'{mds_dtype} is not supported by MDSwriter')
+    return schema_dict
+
+
+def do_merge_index(partitions: Iterable,
+                   mds_path: Union[str, Tuple[str, str]]):
+    """Merge index.json from partitions into one for streaming.
+
+    Args:
+        partitions (Iterable): partitions that contain pd.DataFrame
+        mds_path (Union[str, Tuple[str, str]]): (str,str)=(local,remote), str = local or remote based on parse_uri(url) result
+    Return:
+        None
+    """
+    if not partitions:
+        logger.warning("No partitions exist, no mergei index")
+        return
+
+    shards = []
+
+    for row in partitions:
+        mds_partition_index = f'{row.mds_path}/{get_index_basename()}'
+        mds_partition_basename = os.path.basename(row.mds_path)
+        obj = json.load(open(mds_partition_index))
+        for i in range(len(obj['shards'])):
+            shard = obj['shards'][i]
+            for key in ['raw_data', 'zip_data']:
```

Comment:
```suggestion
            for key in ('raw_data', 'zip_data'):
```
Tuples are faster to construct